### PR TITLE
Session Manager, Enemy Waves, Loot Drops 

### PR DESCRIPTION
**Pull request by rob1997**

### CODE REVIEW

File: Assets/Scripts/Session/SessionManager.cs [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+        protected override void Init()
+        {
+            base.Init();
+            
+            EventSystem<OnSessionStateChanged>.Subscribe(args =>
+            {
+                switch (args.State)
+                {
+                    case SessionState.Playing:
+                        DirectorController = FindObjectOfType<DirectorController>();
```
Comment (kantagara) [inline Assets/Scripts/Session/SessionManager.cs:42]: Couldn't these be singletons, since they're controllers?

**Review by kantagara — APPROVED** [review]
Looks good, small comment left but overall, cool!

File: Assets/Scripts/Session/SessionManager.cs [hunk 2]
```diff
@@ -0,0 +1,69 @@
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using Chainsafe.BlockBlasterz.Events;
+using Chainsafe.BlockBlasterz.Managers;
+using UnityEngine;
+
+namespace Chainsafe.BlockBlasterz
+{
+    public enum SessionState
```
[the remark this answers is not in the document]
Comment (oleksandrchainsafe) [reply Assets/Scripts/Session/SessionManager.cs:10]: All the other stuff like loading level, etc.. are parts of the metagame, and in our case can be handled by simply changing scenes.

File: Assets/Scripts/Session/SessionManager.cs (hunk 1 above)
Comment (rob1997) [reply Assets/Scripts/Session/SessionManager.cs]: done

File: Assets/Prefabs/Weapons/MediumHealthPIckup.prefab [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+  m_PrefabAsset: {fileID: 0}
+  serializedVersion: 6
+  m_Component:
+  - component: {fileID: 9174324197881312602}
+  - component: {fileID: 647582412406466816}
+  - component: {fileID: 5665189927800355556}
+  - component: {fileID: 7858304706744121707}
+  m_Layer: 0
+  m_Name: MediumHealthPIckup
+  m_TagString: Ammo
```
Comment (kantagara) [inline Assets/Prefabs/Weapons/MediumHealthPIckup.prefab:17]: Shouldn't this have a Pickup tag instead of Ammo, Ammo shouldn't be a tag at all imo.

Comment (rob1997) [reply Assets/Prefabs/Weapons/MediumHealthPIckup.prefab:17]: yeah I didn't add it, I duplicated ammo pickup to start that's why it's there

File: Assets/Scripts/Challenges/Reward/EnemyDestroyedReward.cs [hunk 4]
```diff
@@ -0,0 +1,17 @@
+using System.Collections;
+using System.Collections.Generic;
+using Chainsafe.BlockBlasterz.Challenges;
+using UnityEngine;
+
+namespace Chainsafe.BlockBlasterz
+{
+    public class EnemyDestroyedReward : Reward
```
[the remark this answers is not in the document]
Comment (rob1997) [reply Assets/Scripts/Challenges/Reward/EnemyDestroyedReward.cs:8]: I thought maybe we could add item pickups here, plus we would probably reward different scored for destroying different enemies, right?

File: Assets/Prefabs/Weapons/MediumHealthPIckup.prefab (hunk 3 above)
Comment (rob1997) [reply Assets/Prefabs/Weapons/MediumHealthPIckup.prefab]: changed tag to 'untagged'

File: Assets/Scripts/Challenges/Reward/EnemyDestroyedReward.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (kantagara) [reply Assets/Scripts/Challenges/Reward/EnemyDestroyedReward.cs:8]: Yeah but this is a reward on when the _challenge_ is completed, not when the Enemy Dies (I think 😁 )

File: Assets/Scripts/Challenges/ChallengeManager.cs [hunk 5]
```diff
@@ -11,7 +11,7 @@ public class ChallengeManager : Singleton<ChallengeManager>
     {
         [SerializeField] private int maxChallenges = 3;
         [field: SerializeField] public List<Challenge> Challenges { get; set; }
-        protected override bool ShouldntBeDestroyedOnLoad => false;
+        protected override bool ShouldNotDestroyedOnLoad => false;
```
Comment (oleksandrchainsafe) [inline Assets/Scripts/Challenges/ChallengeManager.cs:14]: mb ShouldNot**Be**DestroyedOnLoad? or NoDestroyOnLoad?

File: Assets/Scripts/Session/SessionController.cs [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+        private List<SessionState> _states = new List<SessionState>();
+
+        public SessionState CurrentState { get; private set; }
+        
+        protected override void Init()
+        {
+            base.Init();
+            
+            // add states for example,
+            // TryAddSessionState<LobbyState>();
```
Comment (oleksandrchainsafe) [inline Assets/Scripts/Session/SessionController.cs:33]: Can't find the actual states

File: Assets/Scripts/Session/DirectorController.cs [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+
+        private int _index;
+
+        protected override bool ShouldNotDestroyedOnLoad => false;
+        
+        private void Start()
+        {
+            Debug.Log($"Wave {_index}");
+            
+            waves[_index].BeginWave();
```
Comment (oleksandrchainsafe) [inline Assets/Scripts/Session/DirectorController.cs:21]: This should be triggered by a session state.

File: Assets/Scripts/Session/Wave.cs [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+            }
+        }
+
+        private void TryCompleteWave(OnEnemyDestroyed onEnemyDestroyed)
+        {
+            if (IsWaveCompleted())
+            {
+                EventSystem<OnWaveCompleted>.Invoke(new OnWaveCompleted(this));
+                
+                DirectorController.Instance.NextWave();
```
Comment (oleksandrchainsafe) [inline Assets/Scripts/Session/Wave.cs:32]: Great overall, but again we should give the responsibility of controlling when each wave starts to the session manager. So basically SessionManager instructs the Director to start spawning fellas and watches (or listens) the current wave to end through the Director's interface.

To better illustrate what I mean imagine 3 states:

- ChilingState: Nothing spawns; when the player triggers the new wave, transitions to FarmingState.
- FarmingState: When enabled tells Director to start wave N; watches wave complete -> ChilingState; watches player die -> GameOverState (same might be true for ChilingState in case there is a static damage dealer on the map)
- GameOverState: triggers GameOverUI to show, saves results/stats/best score.

Dependencies can look something like this: SessionManager->Director->Wave. So SessionManager knows nothing about waves except for some basic info like wave number. Director knows nothing of SessionManager, but provides an interface to start/stop a wave with a certain number and to listen for the current wave completed. Wave knows nothing about both of them.
_Basically it's good when a component like director can be moved to another scene where there is no SessionManager and the game wouldn't break._

**Review by oleksandrchainsafe — CHANGES_REQUESTED** [review]

**Review by sneakzttv — APPROVED** [review]

File: Assets/Scripts/Challenges/Reward/EnemyDestroyedReward.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (rob1997) [reply Assets/Scripts/Challenges/Reward/EnemyDestroyedReward.cs]: removed this reward

File: Assets/Scripts/Challenges/ChallengeManager.cs (hunk 5 above)
Comment (rob1997) [reply Assets/Scripts/Challenges/ChallengeManager.cs]: changed to `NoDestroyOnLoad`

File: Assets/Scripts/Session/SessionController.cs (hunk 6 above)
Comment (rob1997) [reply Assets/Scripts/Session/SessionController.cs]: Chilling, farming, paused and game over states implemented

File: Assets/Scripts/Session/DirectorController.cs (hunk 7 above)
Comment (rob1997) [reply Assets/Scripts/Session/DirectorController.cs]: it's not being triggered by farming state

File: Assets/Scripts/Session/GameOverState.cs [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+        public override void Initialize(SessionController sessionController)
+        {
+            base.Initialize(sessionController);
+            
+            EventSystem<OnPlayerDestroyed>.Subscribe(EndGame);
+        }
+
+        private void EndGame(OnPlayerDestroyed onPlayerDestroyed)
+        {
+            SessionController.Instance.TryChangeSessionState<GameOverState>();
```
Comment (oleksandrchainsafe) [inline Assets/Scripts/Session/GameOverState.cs:20]: This part is a bit counterintuitive. I would move this to the Farming & Chilling states.
The problem is that when you read the code, you'd look into the current active state to understand the behavior. With this one the state change happens "out of nowhere", the one debugging this is going to spend some time looking for the reason why the state was changed even though there is no such logic in the current state.

**Review by oleksandrchainsafe — CHANGES_REQUESTED** [review]
All good. The only concern is the state change in GameOverState.

**Review by oleksandrchainsafe — APPROVED** [review]
